### PR TITLE
Spark 2.2 support - update code for spark 2.2  

### PR DESCRIPTION
**Pull request by carsonwang**
## What changes were proposed in this pull request?
Update code for spark 2.2

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
 sudo: required
 dist: trusty
 language: java
-script: mvn test
+script: mvn -Pspark-2.2 test
```
Comment (gczsjdy) [inline .travis.yml]: Don't forget to update to 2 scripts for both versions : )

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapStrategies.scala [hunk 2]
```diff
@@ -214,7 +217,8 @@ trait OapStrategies extends Logging {
         }
 
         val aggregateOperator =
-          if (aggregateExpressions.map(_.aggregateFunction).exists(!_.supportsPartial)) {
+          if (aggregateExpressions.map(_.aggregateFunction).exists(
+            AggregateFunctionAdapter.supportsPartial)) {
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapStrategies.scala]: `!AggregateFunctionAdapter.supportsPartial` ?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 3]
```diff
@@ -187,7 +187,7 @@ case class CreateIndexCommand(
       outputSpec = FileFormatWriter.OutputSpec(
         qualifiedOutputPath.toUri.getPath, Map.empty),
       hadoopConf = configuration,
-      partitionColumns = Seq.empty,
+      Seq.empty, // partitionColumns
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: why we do this change?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapStrategies.scala (hunk 2 above)
Comment (carsonwang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapStrategies.scala]: Thanks for catching this!

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala (hunk 3 above)
Comment (carsonwang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: The interface `FileFormatWriter.write` changed. In spark 2.1, it is `partitionColumns: Seq[Attribute]`, but in Spark 2.2 it is `partitionColumnName: Seq[String]`.

**Comment by LuciferYang** [comment]
@carsonwang I think we should override configuration of maven-compiler-plugin in spark-2.2 profile because spark-2.2 compile by jdk 1.8.

**Comment by LuciferYang** [comment]
@carsonwang refer to SPARK-10364, at line 104 of FileSourceStrategy, parquetOptions need append 
`SQLConf.PARQUET_INT64_AS_TIMESTAMP_MILLIS.key ->_fsRelation.sparkSession.sessionState.conf
  .isParquetINT64AsTimestampMillis.toString` to support this features

File: .travis.yml (hunk 1 above)
Comment (carsonwang) [reply .travis.yml]: Yes, just updated to test both Spark 2.1 and 2.2

**Comment by LuciferYang** [comment]
@carsonwang I found that both the compilation and the unit test were passed, should we merge the code as soon as possible and start integrated test process?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapStrategies.scala [hunk 4]
```diff
@@ -147,8 +151,8 @@ trait OapStrategies extends Logging {
       // We can take a much larger threshold here since if this optimization
       // is applicable, only distinct item will be broadcasted, those data
       // should much less than the origin table.
-      plan.statistics.isBroadcastable ||
-        plan.statistics.sizeInBytes <= conf.autoBroadcastJoinThreshold
+      val stats = LogicalPlanAdapter.getStatistics(plan, conf)
+      stats.sizeInBytes <= conf.autoBroadcastJoinThreshold
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapStrategies.scala]: `stats.isBroadcastable || ...`?

File: src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaSuite.scala [hunk 5]
```diff
@@ -368,87 +368,87 @@ class ParquetSchemaSuite extends ParquetSchemaTest {
...
-        StructType(Seq(
-          StructField("firstfield", StringType, nullable = true),
-          StructField("secondfield", StringType, nullable = true),
-          StructField("thirdfield", StringType, nullable = false))),
-        StructType(Seq(
-          StructField("firstField", StringType, nullable = true),
-          StructField("secondField", StringType, nullable = true))))
-    }.getMessage.contains("detected conflicting schemas"))
-  }
+//  test("merge with metastore schema") {
```
Comment (gczsjdy) [inline src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaSuite.scala]: ?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala [hunk 6]
```diff
@@ -494,8 +494,7 @@ private[oap] class FilterOptimizer(keySchema: StructType) {
       interval1 <- intervalArray1
       interval2 <- intervalArray2
       // isNull & otherPredicate => empty
-      if (!(interval1.isNullPredicate ^ interval2.isNullPredicate)) &&
-        !interval1.isPrefixMatch && !interval2.isPrefixMatch
+      if (!(interval1.isNullPredicate ^ interval2.isNullPredicate))
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: cc @adrian-wang , in spark 2.2, the unit test "startswith using index" failed. The reason is that both two filters isNotNull(b) and startswith(this3) are pushed down. So when these two intervals are merged in this function, the merge result is not correct. But in spark 2.1, only startswith(this3) is pushed down so no merging happened. I am not sure what caused the difference and also the code change here.

File: src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaSuite.scala (hunk 5 above)
Comment (carsonwang) [reply src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaSuite.scala]: The API `ParquetFileFormat.mergeMetastoreParquetSchema` doesn't exist in Spark 2.2. So I disabled these tests temporarily.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapStrategies.scala (hunk 4 above)
Comment (carsonwang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapStrategies.scala]: `stats.isBroadcastable` doesn't exist in Spark 2.2. `stats.sizeInBytes <= conf.autoBroadcastJoinThreshold` should be enough to express the same meaning.

File: src/main/spark2.2/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 7]
```diff
@@ -0,0 +1,1079 @@
...
+    | STATISTICS | ANALYZE | PARTITIONED | EXTERNAL | DEFINED | RECORDWRITER
+    | REVOKE | GRANT | LOCK | UNLOCK | MSCK | REPAIR | RECOVER | EXPORT | IMPORT | LOAD | VALUES | COMMENT | ROLE
+    | ROLES | COMPACTIONS | PRINCIPALS | TRANSACTIONS | INDEX | INDEXES | LOCKS | OPTION | LOCAL | INPATH
+    | ASC | DESC | LIMIT | RENAME | SETS
+    | AT | NULLS | OVERWRITE | ALL | ALTER | AS | BETWEEN | BY | CREATE | DELETE
+    | DESCRIBE | DROP | EXISTS | FALSE | FOR | GROUP | IN | INSERT | INTO | IS |LIKE
+    | NULL | ORDER | OUTER | TABLE | TRUE | WITH | RLIKE
+    | AND | CASE | CAST | DISTINCT | DIV | ELSE | END | FUNCTION | INTERVAL | MACRO | OR | STRATIFY | THEN
+    | UNBOUNDED | WHEN
+    | DATABASE | SELECT | FROM | WHERE | HAVING | TO | TABLE | WITH | NOT | CURRENT_DATE | CURRENT_TIMESTAMP
```
Comment (zhixingheyi-tian) [inline src/main/spark2.2/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: @carsonwang 
In the non-terminal "nonReserved", it seems that the "CHECK", "DISABLE", "ENABLE" terminal symbols are missing. 
from Xiangxiang

**Comment by carsonwang** [comment]
@zhixingheyi-tian , it looks like you were reviewing a previous commit.

File: src/main/spark2.2/scala/org/apache/spark/sql/hive/thriftserver/OapEnv.scala [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SQLContext
+import org.apache.spark.sql.hive.{HiveExternalCatalog, HiveUtils}
+import org.apache.spark.sql.oap.OapSession
+import org.apache.spark.sql.oap.listener.OapListener
+import org.apache.spark.sql.oap.ui.OapTab
+import org.apache.spark.util.Utils
+
+private[hive] object OapEnv extends Logging {
```
[the remark this answers is not in the document]
Comment (carsonwang) [reply src/main/spark2.2/scala/org/apache/spark/sql/hive/thriftserver/OapEnv.scala]: good point. updated.

**Comment by carsonwang** [comment]
Updated readme

File: src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaSuite.scala (hunk 5 above)
Comment (gczsjdy) [reply src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaSuite.scala]: Seems `HiveMetastoreCatalog.mergeWithMetastoreSchema` can replace this function, see https://github.com/apache/spark/pull/16944

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala (hunk 6 above)
Comment (gczsjdy) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: cc @linhong-intel

File: src/main/spark2.2/scala/org/apache/spark/sql/hive/thriftserver/OapEnv.scala [hunk 9]
```diff
@@ -28,6 +28,10 @@ import org.apache.spark.sql.oap.listener.OapListener
 import org.apache.spark.sql.oap.ui.OapTab
 import org.apache.spark.util.Utils
 
+/**
+ * Most of the code in init() are copied from SparkSQLEnv. Please include code from the
+ * corresponding Spark version.
+ */
```
Comment (gczsjdy) [inline src/main/spark2.2/scala/org/apache/spark/sql/hive/thriftserver/OapEnv.scala]: Good! : )

File: src/main/spark2.2/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java [hunk 10]
```diff
@@ -145,7 +150,7 @@ public InternalRow copy() {
           } else if (dt instanceof LongType) {
             row.setLong(i, getLong(i));
           } else if (dt instanceof FloatType) {
-              row.setFloat(i, getFloat(i));
+            row.setFloat(i, getFloat(i));
           } else if (dt instanceof DoubleType) {
```
Comment (zhixingheyi-tian) [inline src/main/spark2.2/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: @carsonwang 
Line 153, there seems to be no change
from Xiangxiang

Comment (carsonwang) [reply src/main/spark2.2/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: This is just to fix the style.

File: src/main/spark2.2/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 11]
```diff
@@ -1368,24 +1392,23 @@ class DAGScheduler(
...
+   * executor serves its own blocks (i.e., we're not using external shuffle) OR a FetchFailed
+   * occurred, in which case we presume all shuffle data related to this executor to be lost.
    *
    * Optionally the epoch during which the failure was caught can be passed to avoid allowing
    * stray fetch failures from possibly retriggering the detection of a node as lost.
    */
   private[scheduler] def handleExecutorLost(
       execId: String,
-      filesLost: Boolean,
+      fetchFailed: Boolean,
```
[the remark this answers is not in the document]
Comment (carsonwang) [reply src/main/spark2.2/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Good catch. This is not necessary. Just copied from Spark 2.1

File: src/main/spark2.2/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+    s"""
+       |if ($batch == null) {
+       |  $nextBatch();
+       |}
+       |while ($batch != null) {
+       |  int $numRows = $batch.numRows();
+       |  int $localEnd = $numRows - $idx;
+       |  for (int $localIdx = 0; $localIdx < $localEnd; $localIdx++) {
+       |    int $rowidx = $idx + $localIdx;
+       |    if ($batch.isFiltered($rowidx)) continue;
```
Comment (zhixingheyi-tian) [inline src/main/spark2.2/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala]: @carsonwang 
What is the purpose of adding this line? Compared to spark 2.2
from Xiangxiang

Comment (LuciferYang) [reply src/main/spark2.2/scala/org/apache/spark/sql/execution/ColumnarBatchScan.scala]: Some recoreds filter by index and we mark it in filteredRows of ColumnarBatch, if `batch.isFiltered(rowidx)` return true, we needn't to push data to downstream.

**Comment by LuciferYang** [comment]
@carsonwang #825(#829 ) will import spark code includes `VectorizedRleValuesReader`, `VectorizedPlainValuesReader `,`VectorizedRleValuesReader ` and change some access modifier

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala (hunk 6 above)
Comment (carsonwang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/IndexContext.scala]: Synced with @adrian-wang , he suggested changing startswith（abc）to >=abc and <abd for startswith. I will create an issue to track this.

**Comment by carsonwang** [comment]
All right, I am going to merge this. Thanks all for the review.

**Comment by gczsjdy** [comment]
Thanks @carsonwang , seems there is new Spark code changes added for 2.2: like `ColumnarBatchScan`, we can add them in issues?

**Comment by carsonwang** [comment]
yes, add #847 , thanks @gczsjdy

**Pull request merged**